src/arrays/arraymath.jl

```diff
@@ -135,35 +130,33 @@ end
 # exponential of dense matrix
 function Base.expm(qarr::AbstractQuMatrix)
     fc = expm(full(rawcoeffs(qarr)))
-    QAT = similar_type(typeof(qarr))
+    QAT = similar_type(qarr)
     return QAT(fc, bases(qarr))
 end
 #Base.expm(ct::CTranspose) = expm(ct.qarr)'
 
-
 ##################
 # Tensor Product #
 ##################
 
 # General tensor product definitions for orthonormal bases
-function tensor{B<:OrthonormalBasis,T1,T2,N}(qarr1::AbstractQuArray{B,T1,N}, qarr2::AbstractQuArray{B,T2,N})
+function tensor{B,T1,T2,N}(qarr1::AbstractQuArray{B,T1,N}, qarr2::AbstractQuArray{B,T2,N})
     tc = kron(coeffs(qarr1), coeffs(qarr2))
-    QAT = similar_type(promote_type(typeof(qarr1), typeof(qarr2))) 
+    QAT = similar_type(qarr1, qarr2)
     return QAT(tc, map(tensor, bases(qarr1), bases(qarr2)))
 end
 
-function tensor{B<:OrthonormalBasis,T1,T2,N}(qarr1::CTranspose{B,T1,N}, qarr2::CTranspose{B,T2,N})
-    return QuArray(kron(rawcoeffs(qarr1), rawcoeffs(qarr2)),
-                   map(tensor, rawbases(qarr1), rawbases(qarr2)))'
-end
+# defined to resolve ambiguity warnings
+tensor{B}(ct1::DualVector{B}, ct2::DualVector{B}) = tensor(ct1.qarr, ct2.qarr)'
+tensor{B}(ct1::CTranspose{B}, ct2::CTranspose{B}) = tensor(ct1.qarr, ct2.qarr)'
 
-function tensor{B<:OrthonormalBasis}(ket::QuVector{B}, bra::DualVector{B})
+function tensor{B}(ket::AbstractQuVector{B}, bra::DualVector{B})
     return QuArray(kron(coeffs(ket), coeffs(bra)),
```

Comment:
It feels like we had this discussion already? Anyways, at least at the moment we don't have anything that needs the factor length as a parameter, so we can just treat it as a value for now and switch back if necessary :-) I think when we introduced this I had in mind that doing partial traces/transpose could profit from it, but IMO this is to vague to stop this PR.